File: SendRecvCollectives/old/comm_pattern.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <math.h>

/*prototypes for functions used to initialize, and check the answers */
int init_array(int *buffer,int size);
int print_errors(char *label,int error,int root);
int check_broadcast(int *buffer,int count,int root);
int check_scatter(int *buffer,int count,int root);
int check_gather(int *buffer,int count,int root);

int broadcast(int *buffer,int count,int root);
int scatter(int *sendbuffer,int count,int *recvbuffer,int root);
int gather(int *sendbuffer,int count,int *recvbuffer,int root);




int broadcast(int *buffer,int count,int root){
    /* Write this function in exercise 5a.  */
  
}



int scatter(int *sendbuffer,int count,int *recvbuffer,int root) {
    /* Write this function in exercise 5b.  */
}


int gather(int *sendbuffer,int count,int *recvbuffer,int root){
    /* Write this function in exercise 5c.  */
   
}


int main(int argc, char *argv[]){
    int rank,nprocs;
    int size,chunk_size;
    int *array,*receive_array;
    int root;
    double t1,t2;
    
    
    MPI_Init(&argc,&argv);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);

    
    //initialize some variables and arrays
    chunk_size=100;
    size=chunk_size*nprocs;
    array=malloc(sizeof(int)*size);
    receive_array=malloc(sizeof(int)*size);

    /*functions should work with any root process*/
    root=nprocs/2;

    // (a), let's try to do the broadcast operation
    
    init_array(array,size);
    broadcast(array,size,root);
    check_broadcast(array,chunk_size,root);

    
    // (b) let's try to do the scatter operation
    init_array(array,size);
    scatter(array,chunk_size,receive_array,root);
    check_scatter(receive_array,chunk_size,root);
    

    // (c) let's try to do the gather operation
    init_array(array,size);
    gather(array,chunk_size,receive_array,root);
    check_gather(receive_array,chunk_size,root);

    MPI_Finalize();

}
int init_array(int *buffer,int size){
    int i,rank,nprocs;
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);   

    for(i=0;i<size;i++)
        buffer[i]=rank+i*nprocs;
}

int check_broadcast(int *buffer,int count,int root){
    int i,rank,nprocs;
    int error;
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);   

    error=0;
    for(i=0;i<count*nprocs;i++)
        error+=buffer[i]!=root+i*nprocs;

    print_errors("broadcast",error,root);
}
int check_scatter(int *buffer,int count,int root){
    int i,rank,nprocs;
    int error;
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);   

    error=0;
    for(i=0;i<count;i++)
        error+=buffer[i]!=root+(rank*count+i)*nprocs;

    print_errors("scatter",error,root);
}

int check_gather(int *buffer,int count,int root){
    int p,i,rank,nprocs;
    int error;
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);   

    error=0;
    if(rank==root)
        for(p=0;p<nprocs;p++)
            for(i=0;i<count;i++)
                error+=buffer[i+p*count]!=p+i*nprocs;

    print_errors("gather",error,root);
}


/*print number of errors in order*/
int print_errors(char *label,int error,int root){    
    int i,rank,nprocs;
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);   
    
    if(rank==0) {
        for(i=0;i<nprocs;i++){
            if(i>0)
                MPI_Recv(&error,1,MPI_INT,i,MPI_ANY_TAG,MPI_COMM_WORLD,MPI_STATUS_IGNORE);

                
            if(error)
                printf("%s: %d errors for rank %d",label,error,i);
            else
                printf("%s: success for rank %d",label,i);            

            if(i==root)
                printf(" (root-process)\n");
            else
                printf("\n");
        }
    }
    else{
        MPI_Send(&error,1,MPI_INT,0,10,MPI_COMM_WORLD);
    }
    return 0;
}

File: SendRecvCollectives/old/comm_pattern.f90
PROGRAM comm_pattern
  IMPLICIT NONE
  INCLUDE 'mpif.h'
  INTEGER,PARAMETER::chunk_size=100
  INTEGER,PARAMETER::root=1

  INTEGER::ierror, myid,ntasks,count
  INTEGER,ALLOCATABLE:: buf(:),recvbuf(:)
  

  CALL MPI_INIT(ierror)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, myid, ierror)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, ntasks, ierror)
  
  ALLOCATE(buf(chunk_size*ntasks))
  ALLOCATE(recvbuf(chunk_size*ntasks))
  
  CALL init_array(buf)
  CALL broadcast(buf,chunk_size,root)
  CALL check_broadcast(buf,root)


  CALL init_array(buf)
  CALL gather(buf,chunk_size,recvbuf,root)
  CALL check_gather(recvbuf,root)

  CALL init_array(buf)
  CALL scatter(buf,chunk_size,recvbuf,root)
  CALL check_scatter(recvbuf,root)
  

  DEALLOCATE(buf)
  DEALLOCATE(recvbuf)
  CALL MPI_FINALIZE(ierror)
CONTAINS


SUBROUTINE broadcast(buffer,count,root)
   ! write this routine 
   INTEGER, DIMENSION(count),INTENT(INOUT):: buffer
   INTEGER, INTENT(IN):: root,count

END SUBROUTINE broadcast

SUBROUTINE scatter(sendbuffer,count,recvbuffer,root)
  ! write this routine 
  INTEGER, DIMENSION(:),INTENT(IN):: sendbuffer
  INTEGER, DIMENSION(:),INTENT(OUT):: recvbuffer
  INTEGER, INTENT(IN):: root,count

END SUBROUTINE scatter

SUBROUTINE gather(sendbuffer,count,recvbuffer,root)
  ! write this routine 
  INTEGER, DIMENSION(:),INTENT(IN):: sendbuffer
  INTEGER, DIMENSION(:),INTENT(OUT):: recvbuffer
  INTEGER, INTENT(IN):: root,count

 END SUBROUTINE gather
 


!below this point some helper routines are defined

SUBROUTINE init_array(buffer)
  INTEGER, DIMENSION(:),INTENT(INOUT):: buffer
  INTEGER::i

  DO i=1,SIZE(buffer)
    buffer(i)=myid+(i-1)*ntasks
  END DO
  
END SUBROUTINE init_array


SUBROUTINE print_errors(label,errorin,root)
  INTEGER,INTENT(IN):: errorin,root
  CHARACTER(LEN=*):: label
  INTEGER::i,error
  IF ( myid .EQ. 0) THEN 
    error=errorin
    DO i = 0,ntasks-1
      IF (i>0) THEN
        CALL MPI_RECV(error,1,MPI_INTEGER,i,MPI_ANY_TAG,MPI_COMM_WORLD,MPI_STATUS_IGNORE,ierror)
      END IF
      IF (error .NE. 0) THEN
        WRITE(*,*) label,":",error," errors for rank ",i
      ELSE
        WRITE(*,*) label,": success for rank ",i
      END IF
    END DO
  ELSE
    CALL MPI_SEND(errorin,1,MPI_INTEGER,0,10,MPI_COMM_WORLD,ierror)
  END IF

END SUBROUTINE print_errors

SUBROUTINE  check_broadcast(buffer,root)
  INTEGER, DIMENSION(:),INTENT(INOUT):: buffer
  INTEGER, INTENT(IN):: root
  INTEGER::i,error  
  error=0
  DO i=1,chunk_size
    IF (buffer(i) .NE. root+(i-1)*ntasks) THEN
      error=error+1
    END IF
  END DO
  CALL print_errors("broadcast",error,root);
END SUBROUTINE check_broadcast

SUBROUTINE check_scatter(buffer,root)
  INTEGER, DIMENSION(:),INTENT(INOUT):: buffer
  INTEGER, INTENT(IN):: root
  INTEGER::i,error
  error=0
  DO i=1,chunk_size
    IF (buffer(i) .NE. root+(myid*chunk_size+(i-1))*ntasks) THEN
      error=error+1
    END IF
  END DO
  CALL print_errors("scatter",error,root);
END SUBROUTINE check_scatter

SUBROUTINE check_gather(buffer,root)
 INTEGER, DIMENSION(:),INTENT(INOUT):: buffer
 INTEGER, INTENT(IN):: root
 INTEGER::p,i,error
 
 error=0
 IF (myid .EQ. root) THEN
   DO p=0,ntasks-1
     DO i=1,chunk_size
       IF (buffer(i+p*chunk_size) .NE. p+(i-1)*ntasks) THEN
         error=error+1
       END IF
     END DO
   END DO
 END IF
 CALL print_errors("gather",error,root);
END SUBROUTINE check_gather

END PROGRAM comm_pattern

